bluepysnap/nexus/factory.py

```diff
@@ -0,0 +1,368 @@
+        """Return the registered resource types.
+
+        Returns:
+            set: Registered resource types.
+        """
+        return set(self._function_registry.keys())
+
+    def get_available_tools(self, resource_type):
+        """Return the available tools for a given resource type.
+
+        Args:
+            resource_type (str): Type of a nexus resource.
+
+        Returns:
+            list: Available tool names.
+        """
+        return list(self._function_registry.get(resource_type, {}))
+
+    def open(self, resource: Resource, tool=None):
+        """Open the resource and return an entity (resource, proxy).
+
+        Args:
+            resource (kgforge.core.Resource): The resource to be opened.
+            tool (str): Name of the tool to open the resource with, or None to use the default tool.
+
+        Returns:
+            Entity: An entity binding the resource and the opener.
+        """
+        return Entity(
+            resource,
+            helper=self._helper,
+            connector=self._connector,
+            opener=partial(self._open_entity, tool=tool),
+        )
+
+    def _open_entity(self, entity, tool=None):
+        """Open the entity and return the associated instance."""
+        types = entity.type  # type or list of types
+        tool_functions = self._get_tool_functions(types)
+        if tool is None:
+            tool, func = first(tool_functions.items())
+            L.info("Using the default tool %s to open %s", tool, types)
+        elif tool in tool_functions:
+            func = tool_functions[tool]
+            L.info("Using the specified tool %s to open %s", tool, types)
+        else:
+            raise RuntimeError(f"Tool {tool} not found for {types}")
+
+        try:
+            if func is open_emodelconfiguration:
+                # TODO: EModelConfiguration in nexus demo/emodel_pipeline
+                # only have the morphology.name, and can't be auto downloaded
+                return func(entity, self._connector)
+            return func(entity)
+        except Exception as ex:
+            raise RuntimeError(f"Unable to open {types}") from ex
+
+    def _get_tool_functions(self, types):
+        """Iterate over types and return the available functions to open the resource."""
+        available_tool_functions = {
+            resource_type: self._function_registry[resource_type]
+            for resource_type in always_iterable(types)
+            if resource_type in self._function_registry
+        }
+        if not available_tool_functions:
+            raise RuntimeError(f"No available tools to open {types}")
+        if not all_equal(available_tool_functions.values()):
+            raise RuntimeError(f"Multiple tools to open {types}")
+        # available_tool_functions contains only one item, or all the values are identical,
+        # so the first value can be returned
+        return first(available_tool_functions.values())
+
+
+def open_circuit_snap(entity):
```

Comment:
let's move these `openers` to another module; the content can be merged w/ the `wrappers` portion, I think - `tools.py` could be the name of it.  Which should also document that people should use that module to add their own tools.